libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt

```diff
@@ -539,80 +532,55 @@ class MapboxRouteLineUtilsTest {
             MapboxRouteLineUtils.getRouteLegTrafficCongestionProvider
         )
 
-        assertEquals(distances.size, result.size)
-        assertEquals(distances.first(), result[1].distanceFromOrigin, 0.0)
+        assertEquals(route.directionsRoute.completeGeometryToPoints().size, result.size)
+        assertEquals(0.016357747710023185, result[1].offset, 0.0)
         assertEquals(result[0].roadClass, roadClasses!!.first())
-        assertEquals(result[2].distanceFromOrigin, distances.subList(0, 2).sum(), 0.0)
-        assertEquals(distancesSum, result.last().distanceFromOrigin, 0.0)
-        assertEquals(LOW_CONGESTION_VALUE, result.last().trafficCongestionIdentifier)
-        assertEquals("service", result.last().roadClass)
-    }
-
-    // The route used here for testing produced an erroneous edge case. The
-    // getRouteLineTrafficExpressionData method was always producing a distanceFromOrigin value
-    // of 0.0 for the beginning of each route leg. This could cause an error when creating the
-    // traffic expression because the distanceFromOrigin value is used to determine the
-    // percentage of distance traveled. These values need to be in ascending order to create a
-    // valid line gradient expression. This error won't occur in single leg routes and will
-    // only occur in multileg routes when there is a traffic congestion change at the first point in
-    // the leg. This is because duplicate traffic congestion values are dropped. The route
-    // used in the test below has a traffic change at the first point in the second leg and
-    // the distance annotation is 0.0 which would have caused an error prior to the fix this
-    // test is checking for.
-    @Test
-    fun getRouteLineTrafficExpressionDataMultiLegRouteWithTrafficChangeAtWaypoint() {
```

Comment:
Duplicate offsets are now allowed to appear in the result of `extractRouteData`, they are filtered out on expression creation level. I moved to a different class to cover this specific case.

Reply:
We should move forward with the existing implementation but is there any value in having duplicates?  An argument could be made that the existing implementation produces accurate data while removing duplicates results in cleaner data if there's no use for the duplicates.

Reply:
My thinking is that with duplicates we represent the actual stops and value changes found in the route, regardless of how that data is later used. The only element that does not accept duplicates (and practically does not need them) is the expression representation, that's why filtering duplicates out on that level makes sense to me.